### PR TITLE
Problem: memiavl don't support ics23 merkle proof generation

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #976 

Solution:
- add the implementaion, some part if ported from iavl library

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/proof.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		Prefix:       prefix,
+	}
+}
+
+// we cannot get the proofInnerNode type, so we need to do the whole path in one function
+func convertInnerOps(path iavl.PathToLeaf) []*ics23.InnerOp {
+	steps := make([]*ics23.InnerOp, 0, len(path))
+
+	// lengthByte is the length prefix prepended to each of the sha256 sub-hashes
+	var lengthByte byte = 0x20
```
Comment (mmsqe) [inline memiavl/proof.go:115]: can we make this to const

File: memiavl/proof_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+			require.NoError(t, err)
+			require.True(t, tree.VerifyMembership(proof, tc.existKey))
+
+			proof, err = tree.GetNonMembershipProof(tc.nonExistKey)
+			require.NoError(t, err)
+			require.True(t, tree.VerifyNonMembership(proof, tc.nonExistKey))
+
+			// test persisted tree
+			require.NoError(t, tree.WriteSnapshot(tmpDir, false))
+			snapshot, err := OpenSnapshot(tmpDir)
```
Comment (mmsqe) [inline memiavl/proof_test.go:44]: miss to close snapshot

File: memiavl/proof.go (hunk 1 above)
Comment (yihuang) [reply memiavl/proof.go:115]: it's copied from iavl library.

File: memiavl/proof_test.go (hunk 2 above)
Comment (yihuang) [reply memiavl/proof_test.go:44]: The `ptree.Close` will close the snapshot.

**Review by mmsqe — APPROVED** [review]

**Pull request merged**